hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/UnassignRegionHandler.java

```diff
@@ -76,75 +79,75 @@ private HRegionServer getServer() {
+          // reportRegionStateTransition, so the HMaster will think the region is online, before
+          // we actually open the region, as reportRegionStateTransition is part of the opening
+          // process.
+          long backoff = this.retryCounter.getBackoffTimeAndIncrementAttempts();
+          LOG.warn("Received CLOSE for {} which is being OPENED; try again after {}ms",
+            encodedRegionNameAsBytes, backoff);
+          rss.getExecutorService().delayedSubmit(this, backoff, TimeUnit.MILLISECONDS);
+        } else {
+          LOG.info("Received CLOSE for {}, which is CLOSING  but not complete yet",
+            encodedRegionNameAsBytes);
+        }
+        return;
       }
-      return;
-    }
-    HRegion region = rs.getRegion(encodedName);
-    if (region == null) {
-      LOG.debug(
-        "Received CLOSE for a region {} which is not online, and we're not opening/closing.",
-        encodedName);
-      rs.getRegionsInTransitionInRS().remove(encodedNameBytes, Boolean.FALSE);
-      return;
-    }
-    String regionName = region.getRegionInfo().getEncodedName();
-    LOG.info("Close {}", regionName);
-    if (region.getCoprocessorHost() != null) {
-      // XXX: The behavior is a bit broken. At master side there is no FAILED_CLOSE state, so if
-      // there are exception thrown from the CP, we can not report the error to master, and if here
-      // we just return without calling reportRegionStateTransition, the TRSP at master side will
-      // hang there for ever. So here if the CP throws an exception out, the only way is to abort
-      // the RS...
-      region.getCoprocessorHost().preClose(abort);
-    }
-    if (region.close(abort) == null) {
-      // XXX: Is this still possible? The old comment says about split, but now split is done at
-      // master side, so...
-      LOG.warn("Can't close region {}, was already closed during close()", regionName);
-      rs.getRegionsInTransitionInRS().remove(encodedNameBytes, Boolean.FALSE);
-      return;
-    }
-    rs.removeRegion(region, destination);
-    if (!rs.reportRegionStateTransition(new RegionStateTransitionContext(TransitionCode.CLOSED,
-      HConstants.NO_SEQNUM, closeProcId, -1, region.getRegionInfo()))) {
-      throw new IOException("Failed to report close to master: " + regionName);
+      HRegion region = rss.getRegion(this.encodedRegionName);
+      // Can't unassign a Region that is not online.
+      if (region == null) {
+        LOG.debug("Received CLOSE for {} which is not online, and we're not opening/closing.",
+          this.encodedRegionName);
+        // Remove what we added above.
+        rss.getRegionsInTransitionInRS().remove(encodedRegionNameAsBytes, Boolean.FALSE);
+        return;
+      }
+      CloseRegionHandler crh = new CloseRegionHandler(rss, region.getRegionInfo(), this.abort,
+          this.destination) {
+        @Override boolean reportRegionStateTransition() throws IOException {
```

Comment:
I find this structure of creating an instance of a sibling class and overriding this method very surprising. I don't have a better suggestion ATM.